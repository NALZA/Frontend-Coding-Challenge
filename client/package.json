{
	"name": "client",
	"version": "0.1.0",
	"private": true,
	"dependencies": {
		"@apollo/client": "^3.3.21",
		"@apollo/react-hooks": "^4.0.0",
		"@apollo/react-testing": "^4.0.0",
		"@testing-library/jest-dom": "^5.14.1",
		"@testing-library/react": "^11.2.7",
		"@testing-library/user-event": "^12.8.3",
		"@types/jest": "^26.0.24",
		"@types/node": "^12.20.16",
		"@types/react": "^17.0.14",
		"@types/react-dom": "^17.0.9",
		"@types/styled-components": "^5.1.11",
		"apollo-client": "^2.6.10",
		"graphql": "^15.5.1",
		"react": "^17.0.2",
		"react-dom": "^17.0.2",
		"react-scripts": "4.0.3",
		"react-test-renderer": "^17.0.2",
		"styled-components": "^5.3.0",
		"typescript": "^4.3.5",
		"web-vitals": "^1.1.2"
	},
	"scripts": {
		"start": "react-scripts start",
		"build": "react-scripts build",
		"test": "react-scripts test --watchAll --coverage",
		"eject": "react-scripts eject"
	},
	"eslintConfig": {
		"extends": [
			"react-app",
			"react-app/jest"
		]
	},
	"browserslist": {
		"production": [
			">0.2%",
			"not dead",
			"not op_mini all"
		],
		"development": [
			"last 1 chrome version",
			"last 1 firefox version",
			"last 1 safari version"
		]
	},
	"devDependencies": {
		"@testing-library/react": "^11.2.7",
		"@types/react-test-renderer": "^17.0.1",
		"jest-styled-components": "^7.0.5"
	},
	"jest": {
		"collectCoverageFrom": [
			"src/**/*.{ts,tsx}",
			"!src/index.tsx",
			"!src/react-app-env.d.ts",
			"!src/graphql/*"
		]
	}
}
