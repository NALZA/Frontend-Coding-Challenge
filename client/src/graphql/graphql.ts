import {
  ApolloClient,
  InMemoryCache,
  HttpLink,
  from,
} from "@apollo/client";
import { onError } from "@apollo/client/link/error";
import awsconfig from "./aws-exports";

export const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.map(({ message, locations, path }) =>
      console.log(
        `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
      )
    );
  if (networkError) console.log(`[Network error]: ${networkError}`);
});

export const link = from([
  errorLink,
  new HttpLink({ uri: awsconfig.aws_appsync_graphqlEndpoint,
  headers: {
    "X-Api-Key": awsconfig.aws_appsync_apiKey
  }
   }),
]);

export const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: link,
});