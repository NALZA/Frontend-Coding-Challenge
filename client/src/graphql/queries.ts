//removed email from query as it is not need for this project=
import { gql } from "@apollo/client";

export const ListZellerCustomers = gql`
  query ListZellerCustomers {
  listZellerCustomers {
    items {
      id
      name
      role
    }
  }
}
`