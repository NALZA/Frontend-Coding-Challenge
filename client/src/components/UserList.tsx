import { useQuery } from '@apollo/client'
import React from 'react'
import { ListZellerCustomers } from '../graphql/queries'
import { UserItem } from './UserItem'

interface Props {
    userType: string,
}

export const UserList = (props: Props) => {
    interface User {
        id: string,
        name: string,
        role: string
    }

    const { loading, data } = useQuery(ListZellerCustomers);

    if (loading) return <p>Loading...</p>;
    if (!data) return <p>Not found</p>;

    const users = 
        data.listZellerCustomers.items.filter((val:User) => {
            return val.role === props.userType
        }).map((val:User) => {
            return <UserItem key={val.id} name={val.name} role={val.role}></UserItem>
    })

    return (
        <div>
            <h2 data-testid="usertype">{props.userType} Users</h2>
            <div>
                {users}
            </div>
        </div>
    )
}
