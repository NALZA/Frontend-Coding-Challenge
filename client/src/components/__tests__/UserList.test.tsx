import { render} from '@testing-library/react';
import { UserList } from '../UserList';
import { MockedProvider } from "@apollo/client/testing";
import { ListZellerCustomers } from '../../graphql/queries';
import TestRenderer from 'react-test-renderer'
import { GraphQLError } from 'graphql';

const mocks = [
    {
      request: {
        query: ListZellerCustomers,
      },
      result: {
        data: {
          listZellerCustomers: {
            "items": [
                {
                    "id": "545eba38-680c-4b7b-9250-3f514146f81a",
                    "name": "John Smith",
                    "role": "Admin"
                },
                {
                    "id": "947a7361-c3ea-4c62-a8c4-649460fc6fcd",
                    "name": "Nathan Smith",
                    "role": "Manager"
                },
                {
                    "id": "066cf6cd-0856-4d9b-98b8-6428bb219239",
                    "name": "Glen Tailor",
                    "role": "Manager"
                }]
          }
        }
      },
    },
  ];

describe("User list holder", () => {
    it("Test graphql loading state", () => {
      const component = TestRenderer.create(
            <MockedProvider mocks={mocks} addTypename={false}>
                <UserList userType='Admin'/>
            </MockedProvider>,  
            );
            const p = component.root.findByType('p');
            expect(p.children.join('')).toContain('Loading...');
    })
    it("Test graphql mock data", async () => {
      const component = render(
            <MockedProvider mocks={mocks} addTypename={false}>
                <UserList userType='Admin'/>
            </MockedProvider>,  
            );
            await new Promise(resolve => setTimeout(resolve, 0));

            const userType = component.getByTestId('usertype');
            expect(userType).toHaveTextContent('Admin Users')
    })
    it("Test graphql no data state", async () => {
      const mocksNodata = [
        {
          request: {
            query: ListZellerCustomers,
          },
          result: {
          },
        },
      ];
      const component = TestRenderer.create(
            <MockedProvider mocks={mocksNodata} addTypename={false}>
                <UserList userType='Admin'/>
            </MockedProvider>,  
            );
            await new Promise(resolve => setTimeout(resolve, 0));
            const p = component.root.findByType('p');
            expect(p.children.join('')).toContain('Not found');
    })
})
