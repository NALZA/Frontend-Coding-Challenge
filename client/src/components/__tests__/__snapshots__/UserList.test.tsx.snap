// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User list holder Test snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 20px;
  height: 60px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: SteelBlue;
  background-color: #e4f3ff;
  border-radius: 10px;
  width: 50px;
  height: 50px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 5px;
}

.c5 {
  margin: 0;
  font-size: .95em;
  color: gray;
}

.c4 {
  margin: 0;
  font-size: 1.17em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 20px;
}

<body>
    <div>
      <div>
        <h2
          data-testid="usertype"
        >
          Admin
           Users
        </h2>
        <div
          class="c0"
        >
          <div
            class="c1"
          >
            <h3
              class="c2"
              data-testid="avatar"
            >
              J
            </h3>
            <div
              class="c3"
            >
              <p
                class="c4"
                data-testid="name"
              >
                John Smith
              </p>
              <p
                class="c5"
                data-testid="role"
              >
                Admin
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2
        data-testid="usertype"
      >
        Admin
         Users
      </h2>
      <div
        class="sc-jSFjdj mBoku"
      >
        <div
          class="sc-bdnxRM khowmD"
        >
          <h3
            class="sc-gtsrHT lanIWJ"
            data-testid="avatar"
          >
            J
          </h3>
          <div
            class="sc-dlnjwi jMwFjE"
          >
            <p
              class="sc-eCApnc ignkqQ"
              data-testid="name"
            >
              John Smith
            </p>
            <p
              class="sc-hKFxyN jvCncI"
              data-testid="role"
            >
              Admin
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
